server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/RecipientRewriteTable.java

```diff
@@ -45,35 +58,78 @@
     }
 
+    @Inject
+    public final void setProcessor(RecipientRewriteTableProcessor processor) {
+        this.processor = processor;
+    }
     /**
-     * @see org.apache.james.transport.mailets.AbstractRecipientRewriteTableMailet#processMail(MailAddress, MailAddress, MimeMessage)
+     * The service rewrite the recipient list of mail. The method should:
+     * - Set Return-Path and remove all other Return-Path headers from the mail's message. This only works because there is a placeholder inserted by MimeMessageWrapper
+     * - If there were errors, we redirect the email to the ERROR processor. In order for this server to meet the requirements of the SMTP
+     * specification, mails on the ERROR processor must be returned to the sender. Note that this email doesn't include any details
+     * regarding the details of the failure(s). In the future we may wish to address this.
+     * - Set the mail's state to <code>Mail.GHOST</code> if the recipients be empty after rewriting.
      */
-    public Collection<MailAddress> processMail(MailAddress sender, MailAddress recipient, MimeMessage message) throws MessagingException {
-        try {
-            Mappings mappings = vut.getMappings(recipient.getLocalPart(), recipient.getDomain());
+    @Override
+    public void service(Mail mail) throws MessagingException {
+        Preconditions.checkNotNull(mail);
+        MimeMessage message = mail.getMessage();
+
+        if (message != null) {
+            Builder<MailAddress> errors = ImmutableList.builder();
+            processor.replaceReturnPath(mail);
+
+            Function<MailAddress, Collection<MailAddress>> processMail = getProcessMailFunction(mail, errors);
+            
+            ImmutableList<MailAddress> newRecipients = processor.rewriteRecipients(mail, processMail);
+
+            if (!errors.build().isEmpty()) {
+                getMailetContext().sendMail(mail.getSender(), errors.build(), message, Mail.ERROR);
+            }
 
-            if (mappings != null) {
-                return handleMappings(mappings, sender, recipient, message);
+            if (newRecipients.isEmpty()) {
+                mail.setState(Mail.GHOST);
             }
-        } catch (ErrorMappingException e) {
-            String errorBuffer = "A problem as occoured trying to alias and forward user " + recipient + ": " + e.getMessage();
-            throw new MessagingException(errorBuffer);
-        } catch (RecipientRewriteTableException e) {
-            throw new MessagingException("Unable to access RecipientRewriteTable", e);
+
+            mail.setRecipients(newRecipients);
         }
 
-        Collection<MailAddress> rcpts = new ArrayList<MailAddress>();
-        rcpts.add(recipient);
-        return rcpts;
     }
 
-    /**
-     * @see org.apache.mailet.base.GenericMailet#getMailetInfo()
-     */
+    private Function<MailAddress, Collection<MailAddress>> getProcessMailFunction(final Mail mail, final Builder<MailAddress> errors) {

```

Comment:
There is a lot of duplication between this and UsersRepositoryAliasingForwarding.java .

What you can do is  :
 - have a method on Processor that takes a MailAddress and returns a Pair of List of MailAddress and error (inner class? Also a isOk() method could be cool).
 - You need to add a RecipientRewriteTable (data-api)  field to your processor.

Thus  wiring will be easier.